File: Retos/Reto #4 - PRIMO, FIBONACCI Y PAR [Media]/python/jaimegamm.py
''' * Escribe un programa que, dado un número, compruebe y muestre si es primo, fibonacci y par.
 * Ejemplos:
 * - Con el número 2, nos dirá: "2 es primo, fibonacci y es par"
 * - Con el número 7, nos dirá: "7 es primo, no es fibonacci y es impar"'''

def es_primo(numero):
    if numero < 2:
        return False
    for i in range(2, int(numero**0.5) + 1):
        if numero % i == 0:
            return False
    return True

def es_fibonacci(numero):
    a, b = 0, 1
    while a <= numero:
        if a == numero:
            return True
        a, b = b, a + b
    return False

def es_par(numero):
    return numero % 2 == 0

def main():
    numero = int(input("Ingrese un número: "))

    if es_primo(numero):
        primo_str = "es primo"
    else:
        primo_str = "no es primo"

    if es_fibonacci(numero):
        fibonacci_str = "es fibonacci"
    else:
        fibonacci_str = "no es fibonacci"

    if es_par(numero):
        par_str = "es par"
    else:
        par_str = "es impar"

    resultado = f"{numero} {primo_str}, {fibonacci_str} y {par_str}"
    print(resultado)
    main()

if __name__ == "__main__":
    main()

File: Retos/Reto #4 - PRIMO, FIBONACCI Y PAR [Media]/javascript/jaimegamm.js
/*# Reto #4: PRIMO, FIBONACCI Y PAR
#### Dificultad: Media | Publicación: 23/01/23 | Corrección: 30/01/23

## Enunciado

 * Escribe un programa que, dado un número, compruebe y muestre si es primo, fibonacci y par.
 * Ejemplos:
 * - Con el número 2, nos dirá: "2 es primo, fibonacci y es par"
 * - Con el número 7, nos dirá: "7 es primo, no es fibonacci y es impar"

*/

function esPrimo(numero) {
    for (let i = 2; i < numero; i++) {
        if (numero % i === 0) {
            return false;
        }
    }
    return numero > 1;
}

function esPar(numero) {
    return numero % 2 === 0;
}

function esFibonacci(numero) {
    const sqrt5 = Math.sqrt(5);
    const phi = (1 + sqrt5) / 2;

    const a = phi * numero;
    return Number.isInteger(a) && a > 0;
}

function analizarNumero(numero) {
    const esPrimoResultado = esPrimo(numero) ? "es primo" : "no es primo";
    const esParResultado = esPar(numero) ? "es par" : "es impar";
    const esFibonacciResultado = esFibonacci(numero) ? "es fibonacci" : "no es fibonacci";

    console.log(`${numero} ${esPrimoResultado}, ${esFibonacciResultado} y ${esParResultado}`);
}

// Ejemplos de uso:
analizarNumero(2);
analizarNumero(7);

